core/unit_test/TestBitManipulationBuiltins.hpp

```diff
@@ -0,0 +1,419 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#include <gtest/gtest.h>
+
+#include <Kokkos_Core.hpp>
+
+// clang-format off
+template <class>
+struct type_helper;
+#define DEFINE_TYPE_NAME(T) \
+template <> struct type_helper<T> { static char const * name() { return #T; } };
+DEFINE_TYPE_NAME(unsigned char)
+DEFINE_TYPE_NAME(unsigned short)
+DEFINE_TYPE_NAME(unsigned int)
+DEFINE_TYPE_NAME(unsigned long)
+DEFINE_TYPE_NAME(unsigned long long)
+#undef DEFINE_TYPE_NAME
+// clang-format on
+
+#define DEFINE_BIT_MANIPULATION_FUNCTION_EVAL(FUNC)   \
+  struct BitManipFunction_##FUNC {                    \
+    template <class T>                                \
+    static KOKKOS_FUNCTION auto eval_constexpr(T x) { \
+      return Kokkos::FUNC(x);                         \
+    }                                                 \
+    template <class T>                                \
+    static KOKKOS_FUNCTION auto eval_builtin(T x) {   \
+      return Kokkos::Experimental::FUNC##_builtin(x); \
+    }                                                 \
+    static char const* name() { return #FUNC; }       \
+  }
+
+DEFINE_BIT_MANIPULATION_FUNCTION_EVAL(countl_zero);
+DEFINE_BIT_MANIPULATION_FUNCTION_EVAL(countl_one);
+DEFINE_BIT_MANIPULATION_FUNCTION_EVAL(countr_zero);
+DEFINE_BIT_MANIPULATION_FUNCTION_EVAL(countr_one);
+DEFINE_BIT_MANIPULATION_FUNCTION_EVAL(popcount);
+DEFINE_BIT_MANIPULATION_FUNCTION_EVAL(has_single_bit);
+DEFINE_BIT_MANIPULATION_FUNCTION_EVAL(bit_ceil);
+DEFINE_BIT_MANIPULATION_FUNCTION_EVAL(bit_floor);
+DEFINE_BIT_MANIPULATION_FUNCTION_EVAL(bit_width);
+
+template <class Space, class Func, class Arg, std::size_t N>
+struct TestBitManipFunction {
+  Arg val_[N];
+  TestBitManipFunction(const Arg (&val)[N]) {
+    std::copy(val, val + N, val_);
```

Comment:
Maybe also make sure here that the `*_builtin` can be used in a `constexpr` context?

Reply:
??  The `Experimental::*_builtin` function templates are not declared `constexpr`.

Reply:
The public functions are callable in a `constexpr` context (TestBitManipulation.cpp specifically tests that).

Given that, what is the use case behind `*_builtin` being directly useable in a `constexpr` context?

Reply:
I missed the test.